File: apps/telegram/ui/callbacks.py
"""Callback naming — mirrors telegram/callbacks.ts."""

from __future__ import annotations

from dataclasses import dataclass

CALLBACK_PREFIX = "pmx:"
LEGACY_PREFIXES = ("act:", "trade:", "brief:", "link:", "queue:", "mode:")


@dataclass(frozen=True)
class ParsedCallback:
    raw: str
    prefix: str
    action: str
    parts: tuple[str, ...]


def build_callback(action: str, *parts: str) -> str:
    body = ":".join([action, *parts]) if parts else action
    data = f"{CALLBACK_PREFIX}{body}"
    if len(data) > 64:
        raise ValueError(f"callback_data too long ({len(data)}): {data}")
    return data


def parse_callback(data: str) -> ParsedCallback:
    if data.startswith(CALLBACK_PREFIX):
        rest = data[len(CALLBACK_PREFIX) :]
        parts = tuple(rest.split(":")) if rest else ()
        action = parts[0] if parts else ""
        return ParsedCallback(raw=data, prefix=CALLBACK_PREFIX, action=action, parts=parts[1:])
    legacy = next((p for p in LEGACY_PREFIXES if data.startswith(p)), "")
    action = legacy.replace(":", "") if legacy else "unknown"
    return ParsedCallback(raw=data, prefix=legacy, action=action, parts=tuple(data.split(":")[1:]))

File: telegram-ui/callbacks.ts
/**
 * Callback data naming for pmxtrader Telegram UI.
 * Prefix `pmx:` keeps pmxtrader callbacks distinct from legacy `act:` / `trade:` handlers.
 */

export const CALLBACK_PREFIX = "pmx:" as const;
export const LEGACY_PREFIXES = ["act:", "trade:", "brief:", "link:", "queue:", "mode:"] as const;

export type CallbackAction =
  | "menu"
  | "nav"
  | "refresh"
  | "toggle"
  | "quick"
  | "confirm"
  | "cancel"
  | "page"
  | "noop";

export interface ParsedCallback {
  raw: string;
  prefix: string;
  action: CallbackAction | string;
  parts: string[];
}

/** Build callback_data string (max 64 bytes — keep segments short). */
export function buildCallback(action: string, ...parts: string[]): string {
  const body = [action, ...parts].filter(Boolean).join(":");
  const data = `${CALLBACK_PREFIX}${body}`;
  if (data.length > 64) {
    throw new Error(`callback_data too long (${data.length}): ${data}`);
  }
  return data;
}

export function parseCallback(data: string): ParsedCallback {
  if (data.startsWith(CALLBACK_PREFIX)) {
    const rest = data.slice(CALLBACK_PREFIX.length);
    const parts = rest.split(":");
    return {
      raw: data,
      prefix: CALLBACK_PREFIX,
      action: parts[0] ?? "",
      parts: parts.slice(1),
    };
  }
  const legacy = LEGACY_PREFIXES.find((p) => data.startsWith(p));
  return {
    raw: data,
    prefix: legacy ?? "",
    action: legacy ? legacy.replace(":", "") : "unknown",
    parts: data.split(":").slice(1),
  };
}

export const QUICK_ACTIONS = ["analyze", "watch", "trade", "alert"] as const;
export type QuickAction = (typeof QUICK_ACTIONS)[number];
